### ISSUE TITLE
Make `urlize` behave like other Hugo URLs do

### ISSUE THREAD

**Issue by Jos512** [gohugoio/hugo#4220]
### Problem

When constructing URLs with [the `urlize` function](https://gohugo.io/functions/urlize/), those URLs can differ from the URLs that Hugo/BlackFriday generated. In my case, I have issues because `urlize` comes up with something else than the `#` elements that are linked to in the TOC.

### Replication

Say I make headlines in my articles like so:

~~~
## Hello & bye

## .NET

## 1/3rd done

## .01 improvement

## Hélène

## 1 + 1 = 3?
~~~

These render the following links in the table of contents:

~~~
<nav id="TableOfContents">
<ul>
    <li>
    <ul>
        <li><a href="#hello-bye">Hello &amp; bye</a></li>
        <li><a href="#net">.NET</a></li>
        <li><a href="#1-3rd-done">1/3rd done</a></li>
        <li><a href="#01-improvement">.01 improvement</a></li>
        <li><a href="#hélène">Hélène</a></li>
        <li><a href="#1-1-3">1 + 1 = 3?</a></li>
    </ul></li>
</ul>
</nav>
~~~

And so as far as the TOC is concerned, the ID anchors are:

~~~
#hello-bye
#net
#1-3rd-done
#01-improvement
#hélène
#1-1-3
~~~

But now let's ask `urlize` to turn those same string values into valid URLs like so:

~~~
{{ urlize "hello & bye" }}<br>
{{ urlize ".NET" }}<br>
{{ urlize "1/3rd done" }}<br>
{{ urlize ".01 improvement" }}<br>
{{ urlize "Hélène" }}<br>
{{ urlize "1 + 1 = 3?" }}<br>
~~~

This outcome becomes:

~~~
hello--bye
.net
1/3rd-done
.01-improvement
h%C3%A9l%C3%A8ne
1-+-1--3
~~~

The problem is that *none* of these match the URLs that the TOC came up with. They are incompatible with each other.

### Expected

I'd expect that the `urlize` function behaves like other URLs in Hugo are generated. It's technically correct that `urlize` makes valid URLs, but those URLs can be wildly different from other URLs Hugo makes. 

This makes working with `urlize` a pain since each incompatibility will need to be addressed manually.

**Comment by bep** [comment]
What you suggest would be a pretty breaking update, and I'm not sure I'm willing to take on that work task even if I agreed it was a good thing.

The one issue you mention above is a mismatch between `urlize` and 

https://godoc.org/github.com/russross/blackfriday#hdr-Sanitized_Anchor_Names

Which I would say is something different than sanitizing the full URL. I think we need `anchorize` or something that uses the implementation above + maybe (if possible) adjust `urlize` to detect any named URLs and fix the anchor part.

**Comment by Jos512** [comment]
> What you suggest would be a pretty breaking update, and I'm not sure I'm willing to take on that work task even if I agreed it was a good thing (and the work being the support, answering questions etc.).

That's true. My use case (using `urlize` to generate anchor IDs) hasn't come up earlier (and `urlize` is at least 3+ years old it seems). So it indeed wouldn't be a good idea to change `urlize` just for that.

And I don't expect you to, even though I wrote my issue from the perspective that `urlize` was/is misbehaving. (I should have formulated that clearer.)

> something different than sanitizing the full URL.

That's true, in my case I actually don't need `urlize` to sanitise the full URL. I use it like this:

~~~
{{ $catName := index .Page.Params.categories 1 }}
<a href="{{ .Site.BaseURL }}articles/#{{ $catName | urlize }}">{{ $catName }}</a>
~~~

> I think we need anchorize or something that uses the implementation above + maybe (if possible) adjust urlize to detect any named URLs and fix the anchor part.

I would be very grateful for such an `anchorize` function because that seems more than enough to generate URLs with anchors that match those already made by BlackFriday for the TOC.

From the way I understand the Godoc page you link to, Blackfriday already comes with a function for that: `SanitizedAnchorName()` (see [here](https://godoc.org/github.com/russross/blackfriday#SanitizedAnchorName)).

In the [gopkg](https://github.com/gohugoio/hugo/blob/master/Gopkg.lock) file for Hugo I see that the project also uses `github.com/shurcooL/sanitized_anchor_name` which also sanitises anchors with the same approach as Blackfriday uses (as mentioned on the Godoc page of Blackfriday). So at least that new function (if you choose to add it) doesn't require an additional dependency to manage.

By the way, for the examples I provided in the issue above the `sanitized_anchor_name` resource does a good job. Here's the Go code I used to test:

~~~go
package main

import (
	"fmt"

	"github.com/shurcooL/sanitized_anchor_name"
)

func main() {
	input := [6]string{
		"Hello & bye",
		".NET",
		"1/3rd done",
		".01 improvement",
		"Hélène",
		"1 + 1 = 3?",
	}

	expected := [6]string{
		"hello-bye",
		"net",
		"1-3rd-done",
		"01-improvement",
		"hélène",
		"1-1-3",
	}

	outcome := [6]string{}

	for i, val := range input {
		sanitised := sanitized_anchor_name.Create(val)

		outcome[i] = sanitised
		fmt.Println(sanitised)
	}

	fmt.Println("Does outcome match?", expected == outcome)
}
~~~

And the outcome:

~~~
hello-bye
net
1-3rd-done
01-improvement
hélène
1-1-3
Does outcome match? true
~~~

(Contributing to Hugo is beyond my skills, but perhaps this helps a little bit.)

**Comment by Jos512** [comment]
Thanks for picking up this enhancement and for scheduling it so quickly. Thank you!
